ui-trackoccupancydiagram/src/components/helpers/drawElements/drawOccupancyZones.ts

```diff
@@ -4,58 +4,197 @@ import {
+    // use a for of loop to be able to break the loop if there are more than 9 zones
+    for (const [sortedIndex, zone] of filteredOccupancyZone.entries()) {
       const arrivalTime = getTimePixel(zone.arrivalTime.getTime());
       const departureTime = getTimePixel(zone.departureTime.getTime());
+      const isThroughTrain = arrivalTime === departureTime;
+
+      // reset the overlap check if the zone is not overlapping
+      if (arrivalTime > lastDepartureTimePixel) {
+        yPosition = OCCUPANCY_ZONE_START;
+        primaryArrivalTimePixel = arrivalTime;
+        primaryDepartureTimePixel = departureTime;
+        lastDepartureTimePixel = departureTime;
+        yOffset = 0;
+        counter = 0;
+      }
+
+      // figure out if the zone is overlapping with any previous one
+      // if so and it's an even index, move it to the bottom, if it's an odd index, move it to the top
+      if (arrivalTime >= primaryArrivalTimePixel && arrivalTime < lastDepartureTimePixel) {
+        if (counter % 2 === 0) {
+          yPosition -= yOffset;
+        } else {
+          yPosition += yOffset;
+        }
+        yOffset += Y_OFFSET_INCREMENT;
+        counter += 1;
+      }
+
+      if (departureTime >= lastDepartureTimePixel) lastDepartureTimePixel = departureTime;
+
+      // draw at least 9 zones and zones texts
+      // if there are more than 9 zones, draw a box with the remaining trains
+      if (counter <= MAX_ZONES) {
+        if (selectedTrainId === zone.id) {
+          ctx.fillStyle = SELECTION_20;
+          ctx.beginPath();
+          ctx.roundRect(
+            arrivalTime - X_BACKGROUND_OFFSET - (isThroughTrain ? X_TROUGHTRAIN_PADDING : 0),
+            yPosition - Y_BACKGROUND_OFFSET,
+            departureTime -
+              arrivalTime +
+              X_BACKGROUND_OFFSET * 2 +
+              (isThroughTrain ? X_TROUGHTRAIN_PADDING * 2 : 0),
+            Y_BACKGROUND_OFFSET * 2,
+            SELECTED_TRAIN_ID_GRADIANT
+          );
+          ctx.fill();
+        }
+
+        ctx.fillStyle = zone.color;
+        ctx.strokeStyle = WHITE_100;
+        ctx.lineJoin = 'round';
+        ctx.lineCap = 'round';
+
+        if (isThroughTrain) {
+          drawThroughTrain({ ctx, arrivalTime });
+        } else {
+          drawDefaultZone({ ctx, arrivalTime, departureTime, yPosition });
+        }
+
+        drawOccupancyZonesTexts({
+          ctx,
+          zone,
+          arrivalTime,
+          departureTime,
+          yPosition,
+          isThroughTrain,
+          selectedTrainId,
+        });
+      } else if (counter === MAX_ZONES + 1) {
+        const textX =
+          primaryArrivalTimePixel +
+          (lastDepartureTimePixel - primaryArrivalTimePixel) / 2 -
+          REMAINING_TRAINS_WIDTH / 2;
+        const textY = OCCUPANCY_ZONE_START - REMAINING_TEXT_OFFSET;
+        const text = `+${filteredOccupancyZone.length - sortedIndex} trains`;
+
+        drawRemainingTrainsBox({ ctx, text, textX, textY });
```

Comment:
If some of the const are purely static, it would be better to define them in the function drawRemainingTrainsBox. Ideally, the drawRemainingTrainsBox should only take in arguments what is useful for it, like this for instance
```suggestion
        drawRemainingTrainsBox(ctx, remainingTrainsNb, xPosition });
```

Reply:
done